content/docs/04-ai-sdk-rsc/02-streaming-react-components.mdx

```diff
@@ -0,0 +1,201 @@
+    The weather in {props.location} is {props.weather}
+  </div>
+);
+
+export async function streamComponent() {
+  const result = await streamUI({
+    model: openai('gpt-4o'),
+    prompt: 'Get the weather for San Francisco',
+    text: ({ content }) => <div>{content}</div>,
+    tools: {
+      getWeather: {
+        description: 'Get the weather for a location',
+        parameters: z.object({
+          location: z.string(),
+        }),
+        generate: async function* ({ location }) {
+          yield <LoadingComponent />;
+          const weather = await getWeather(location);
+          return <WeatherComponent weather={weather} location={location} />;
+        },
+      },
+    },
+  });
+
+  return result.value;
+}
+```
+
+The `getWeather` tool should look familiar as it is identical to the example in the previous section. In order for this tool to work:
+
+1. First define a `LoadingComponent`, which renders a pulsing `div` that will show some loading text.
+2. Next, define a `getWeather` function that will timeout for 2 seconds (to simulate fetching the weather externally) before returning the "weather" for a `location`. Note: you could run any asynchronous TypeScript code here.
+3. Finally, define a `WeatherComponent` which takes in `location` and `weather` as props, which are then rendered within a `div`.
+
+Your Server Action is an asynchronous function called `streamComponent` that takes no inputs, and returns a ReactNode. Within the action, you call the `streamUI` function, specifying the model (`gpt-4o`), the prompt, the component that should be rendered if the model chooses to return text, and finally, your `getWeather` tool. Last but not least, you return the resulting component generated by the model with `result.value`.
+
+To call this Server Action and display the resulting React Component, you will need a page.
+
+### Step 2: Create a Page
+
+Create or update your root page (`app/page.tsx`) with the following code:
+
+```tsx filename="app/page.tsx"
+'use client';
+
+import { useState } from 'react';
+import { Button } from '@/components/ui/button';
+import { streamComponent } from './actions';
+
+export default function Page() {
+  const [component, setComponent] = useState<React.ReactNode>();
+
+  return (
+    <div>
+      <form
+        onSubmit={async e => {
+          e.preventDefault();
+          setComponent(await streamComponent());
+        }}
+      >
+        <Button>Stream Component</Button>
+      </form>
+      <div>{component}</div>
+    </div>
+  );
+}
+```
+
+This page is first marked as a client component with the `"use client";` directive given it will be using hooks and interactivity. On the page, you render a form. When that form is submitted, you call the `streamComponent` action created in the previous step (just like any other function). The `streamComponent` action returns a ReactNode that you can then render on the page using React state (`setComponent`).
```

Comment:
```suggestion
This page is first marked as a client component with the `"use client";` directive given it will be using hooks and interactivity. On the page, you render a form. When that form is submitted, you call the `streamComponent` action created in the previous step (just like any other function). The `streamComponent` action returns a `ReactNode` that you can then render on the page using React state (`setComponent`).
```